README.md
# express-compiless

[![NPM version](https://badge.fury.io/js/express-compiless.svg)](http://badge.fury.io/js/express-compiless)
[![Build Status](https://travis-ci.org/papandreou/express-compiless.svg?branch=master)](https://travis-ci.org/papandreou/express-compiless)
[![Coverage Status](https://coveralls.io/repos/papandreou/express-compiless/badge.svg)](https://coveralls.io/r/papandreou/express-compiless)
[![Dependency Status](https://david-dm.org/papandreou/express-compiless.svg)](https://david-dm.org/papandreou/express-compiless)

Middleware that compiles less to css on-the-fly. Intended to be used
in a development setting with the `express.static` middleware, but
should work with any middleware further down the stack, even an http
proxy.

The response will be rewritten under these circumstances:

- If the response is served with a `Content-Type` of `text/less`.
- If the request url ends in `.less` (excluding GET parameters) and
  the `Content-Type` is `application/octet-stream` (this is what
  <a href="https://github.com/senchalabs/connect">Connect</a>'s `static`
  middleware does.

Compiless plays nice with conditional GET. If the original response
has an ETag, compiless will add to it so the ETag of the compiled
response never clashes with the original ETag. That prevents the
middleware issuing the original response from being confused into
sending a false positive `304 Not Modified` if compiless is turned
off or removed from the stack later.

## Installation

Make sure you have node.js and npm installed, then run:

    npm install express-compiless

## Example usage

```javascript
var express = require('express'),
  compiless = require('express-compiless'),
  root = '/path/to/my/static/files';

express()
  .use(compiless({ root: root }))
  .use(express.static(root))
  .listen(1337);
```

## Releases

[Changelog](https://github.com/papandreou/express-compiless/blob/master/CHANGELOG.md)

## License

3-clause BSD license -- see the `LICENSE` file for details.


CHANGELOG.md
### v5.0.0 (2020-11-22)

#### Pull requests

- [#50](https://github.com/papandreou/express-compiless/pull/50) Upgrade prettier to version 2.2.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#48](https://github.com/papandreou/express-compiless/pull/48) Upgrade eslint-config-standard to version 16.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#47](https://github.com/papandreou/express-compiless/pull/47) Upgrade eslint-config-standard to version 15.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))

#### Commits to master

- [Also check \*.json with prettier](https://github.com/papandreou/express-compiless/commit/1d55fd5d24f38b58dda8790761de447f8c4cabb1) ([Andreas Lind](mailto:andreaslindpetersen@gmail.com))
- [Add CHANGELOG.md + offline-github-changelog integration](https://github.com/papandreou/express-compiless/commit/9501e25e9140742ba46f9310bee5bfcd91f36e43) ([Andreas Lind](mailto:andreaslindpetersen@gmail.com))
- [Drop node.js 8 support, add 14, lint with 14 \(semver-major\)](https://github.com/papandreou/express-compiless/commit/3f5ed504d1d8fbef503e72a5e5aa1d9de6997727) ([Andreas Lind](mailto:andreaslindpetersen@gmail.com))
- [Update unexpected and unexpected-express](https://github.com/papandreou/express-compiless/commit/d69ac0eee690f9590d5fa6fe64d104332f2a9bf9) ([Andreas Lind](mailto:andreaslindpetersen@gmail.com))

### v4.2.0 (2020-09-28)

#### Pull requests

- [#46](https://github.com/papandreou/express-compiless/pull/46) Upgrade to hijackresponse 5 ([Gustav Nikolaj Olsen](mailto:gno@one.com))
- [#45](https://github.com/papandreou/express-compiless/pull/45) Upgrade prettier to version 2.1.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#44](https://github.com/papandreou/express-compiless/pull/44) Upgrade eslint-plugin-mocha to version 8.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#41](https://github.com/papandreou/express-compiless/pull/41) Upgrade eslint-plugin-mocha to version 7.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#40](https://github.com/papandreou/express-compiless/pull/40) Upgrade eslint to version 7.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#37](https://github.com/papandreou/express-compiless/pull/37) Upgrade mocha to version 7.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#36](https://github.com/papandreou/express-compiless/pull/36) Upgrade eslint-plugin-node to version 11.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#35](https://github.com/papandreou/express-compiless/pull/35) Upgrade nyc to version 15.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#34](https://github.com/papandreou/express-compiless/pull/34) Upgrade eslint-plugin-node to version 10.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#33](https://github.com/papandreou/express-compiless/pull/33) Upgrade eslint-config-standard to version 14.0.0 ([Andreas Lind](mailto:andreaslindpetersen@gmail.com), [depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#32](https://github.com/papandreou/express-compiless/pull/32) Upgrade eslint-plugin-mocha to version 6.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#31](https://github.com/papandreou/express-compiless/pull/31) Upgrade eslint-config-standard to version 13.0.1 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#30](https://github.com/papandreou/express-compiless/pull/30) Upgrade eslint-config-prettier to version 6.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#29](https://github.com/papandreou/express-compiless/pull/29) Upgrade unexpected-express to version 12.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#27](https://github.com/papandreou/express-compiless/pull/27) Upgrade eslint-config-prettier to version 5.0.0 ([depfu[bot]](mailto:depfu[bot]@users.noreply.github.com))
- [#26](https://github.com/papandreou/express-compiless/pull/26) Upgrade async to version 3.0.0 ([depfu[bot]](mailto:depfu[bot]@users.noreply.github.com))
- [#24](https://github.com/papandreou/express-compiless/pull/24) Upgrade nyc to version 14.0.0 ([depfu[bot]](mailto:depfu[bot]@users.noreply.github.com))

#### Commits to master

- [Upgrade less to ^3.12.2](https://github.com/papandreou/express-compiless/commit/0892630ee24745dafe19fb57b9b3864812d170aa) ([Gustav Nikolaj Olsen](mailto:gno@one.com))
- [Update CI setup, run on node latest, only lint on node 12](https://github.com/papandreou/express-compiless/commit/97bb95d6eaa24b7aea5e41864be86ae8d787f097) ([Andreas Lind](mailto:andreas.lind@peakon.com))
- [prettier --write '\*\*\/\*.{js,md}'](https://github.com/papandreou/express-compiless/commit/a59e628cc5bb158d07bc1bdc26a24d570c97586d) ([Andreas Lind](mailto:andreas.lind@peakon.com))
- [Also enable prettier for markdown files](https://github.com/papandreou/express-compiless/commit/52529c0bb54385a318e073a83a6a9a5aa2f1016c) ([Andreas Lind](mailto:andreas.lind@peakon.com))
- [Update prettier to version 2.0.2](https://github.com/papandreou/express-compiless/commit/b943ddda092f1e4ae55f762a3dcac4fce25e1862) ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [+6 more](https://github.com/papandreou/express-compiless/compare/v4.1.0...v4.2.0)

### v4.1.0 (2019-03-05)

#### Pull requests

- [#23](https://github.com/papandreou/express-compiless/pull/23) Upgrade hijackresponse to version 4.0.0 ([depfu[bot]](mailto:depfu[bot]@users.noreply.github.com))
- [#22](https://github.com/papandreou/express-compiless/pull/22) Upgrade mocha to version 6.0.0 ([depfu[bot]](mailto:depfu[bot]@users.noreply.github.com))
- [#21](https://github.com/papandreou/express-compiless/pull/21) Upgrade async to version 2.6.2 ([depfu[bot]](mailto:depfu[bot]@users.noreply.github.com))

test/compiless.js
const express = require('express');
const Path = require('path');
const unexpected = require('unexpected');
const compiless = require('../lib/compiless');

describe('compiless', () => {
  const root = Path.resolve(__dirname, 'root');
  const expect = unexpected
    .clone()
    .installPlugin(require('unexpected-express'))
    .addAssertion('to yield response', (expect, subject, value) =>
      expect(
        express()
          .use(compiless({ root }))
          .use('/hello', (req, res, next) => {
            res.setHeader('Content-Type', 'text/plain');
            res.setHeader('ETag', 'W/"fake-etag"');
            res.status(200);
            res.write('world');
            res.end();
          })
          .use(express.static(root)),
        'to yield exchange',
        {
          request: subject,
          response: value,
        }
      )
    );

  it('should not mess with request for non-less file', () =>
    expect('GET /something.txt', 'to yield response', {
      headers: {
        'Content-Type': 'text/plain; charset=UTF-8',
        ETag: expect.it('not to match', /-compiless/),
        'Content-Length': '4',
      },
      body: 'foo\n',
    }));

  it('should not mess with request for non-less related route', () =>
    expect('GET /hello', 'to yield response', {
      headers: {
        'Content-Type': 'text/plain',
        ETag: expect.it('not to match', /-compiless/),
      },
      body: 'world',
    }));

  it('should respond with an ETag header and support conditional GET', () =>
    expect('GET /simple.less', 'to yield response', {
      statusCode: 200,
      headers: {
        'Content-Type': 'text/css; charset=utf-8',
        ETag: /^W\/".*-compiless"$/,
      },
      body: '#foo #bar {\n  color: blue;\n}\n',
    }).then((context) => {
      const etag = context.httpResponse.headers.get('ETag');
      return expect(
        {
          url: 'GET /simple.less',
          headers: {
            'If-None-Match': etag,
          },
        },
        'to yield response',
        {
          statusCode: 304,
          headers: {
            ETag: etag,
          },
        }
      );
    }));

  it('should not destroy if-none-match for non .less files', () =>
    expect('GET /something.txt', 'to yield response', {
      statusCode: 200,
      headers: {
        ETag: /^W\/"[a-z0-9-]+"$/,
      },
      body: 'foo\n',
    }).then((context) => {
      const etag = context.httpResponse.headers.get('ETag');
      return expect(
        {
          url: 'GET /something.txt',
          headers: {
            'If-None-Match': etag,
          },
        },
        'to yield response',
        {
          statusCode: 304,
          headers: {
            ETag: etag,
          },
        }
      );
    }));

  it('should compile less file with @import to css with .compilessinclude rules first', () =>
    expect('GET /stylesheet.less', 'to yield response', {
      headers: {
        'Content-Type': 'text/css; charset=utf-8',
      },
      body:
        '.compilessinclude {background-image: url(imports/a.less); display: none;}\nbody {\n  width: 100%;\n}\n#foo #bar {\n  color: red;\n}\n/* multi-line\n   comment\n*/\n',
    }));

  it('should render less file that has a syntax error with the error message as the first thing in the output, wrapped in a body:before rule', () =>
    expect('GET /syntaxerror.less', 'to yield response', {
      statusCode: 200,
      headers: {
        'Content-Type': 'text/css; charset=utf-8',
      },
      body: /Error compiling/,
    }));

  it('should render less file that has an @import error with the error message as the first thing in the output, wrapped in a body:before rule', () =>
    expect('GET /importerror.less', 'to yield response', {
      statusCode: 200,
      headers: {
        'Content-Type': 'text/css; charset=utf-8',
      },
      body: /Error compiling/,
    }));

  it('should render less file that has an @import that points at a file with a syntax error', () =>
    expect('GET /importedsyntaxerror.less', 'to yield response', {
      statusCode: 200,
      headers: {
        'Content-Type': 'text/css; charset=utf-8',
      },
      body: /Error compiling/,
    }));

  it('should render less file that has an second level @import error with the error message as the first thing in the output, wrapped in a body:before rule', () =>
    expect('GET /secondlevelimporterror.less', 'to yield response', {
      statusCode: 200,
      headers: {
        'Content-Type': 'text/css; charset=utf-8',
      },
      body: /Error compiling/,
    }));

  it('should rewrite urls correctly', () =>
    expect(
      'GET /importLessWithRelativeImageReferenceInDifferentDir.less',
      'to yield response',
      {
        statusCode: 200,
        headers: {
          'Content-Type': 'text/css; charset=utf-8',
        },
        body: /url\(imports\/images\/foo.png\)/,
      }
    ));

  it('should deliver a response even though the less file has @imports and references an undefined variable', () =>
    expect('GET /undefinedVariable.less', 'to yield response', {
      statusCode: 200,
      headers: {
        'Content-Type': 'text/css; charset=utf-8',
      },
      body: /Error compiling.*variable.*undefined/,
    }));
});


test/flattenLessText.js
var express = require('express'),
    Path = require('path'),
    request = require('request'),
    passError = require('passerror'),
    expect = require('expect.js'),
    fs = require('fs'),
    flattenLessText = require('../lib/flattenLessText');

describe('flattenLessText', function () {
    it('should rewrite urls correctly', function (done) {
        var baseDir = Path.resolve(__dirname, 'root'),
            lessText = fs.readFileSync(Path.resolve(baseDir, 'importLessWithRelativeImageReferenceInDifferentDir.less'), 'utf-8');

        flattenLessText(lessText, baseDir, baseDir, {}, [], passError(done, function (flattenedLessText) {
            expect(flattenedLessText).to.match(/url\(imports\/images\/foo.png\)/);
            done();
        }));
    });
});


lib/index.js
var Path = require('path'),
    crypto = require('crypto'),
    fs = require('fs'),
    _ = require('underscore'),
    less = require('less'),
    seq = require('seq'),
    flattenLessText = require('./flattenLessText');

require('express-hijackresponse');
require('bufferjs');

module.exports = function compiless(options) {
    if (!options || !options.root) {
        throw new Error('options.root is mandatory');
    }
    return function (req, res, next) {
        if (req.accepts('text/css')) {
            // Prevent If-None-Match revalidation with the downstream middleware with ETags that aren't suffixed with "-compiless":
            var ifNoneMatch = req.headers['if-none-match'];
            if (ifNoneMatch) {
                var validIfNoneMatchTokens = ifNoneMatch.split(" ").filter(function (etag) {
                    return /-compiless\"$/.test(etag);
                });
                if (validIfNoneMatchTokens.length > 0) {
                    req.headers['if-none-match'] = validIfNoneMatchTokens.join(" ");
                } else {
                    delete req.headers['if-none-match'];
                }
            }
            delete req.headers['if-modified-since']; // Prevent false positive conditional GETs after enabling compiless
            res.hijack(function (err, res) {
                var contentType = res.getHeader('Content-Type'),
                    matchContentType = contentType && contentType.match(/^text\/less(?:;\s*charset=([a-z0-9\-]+))?$/i);
                // The mime module doesn't support less yet, so we fall back:
                if (matchContentType || (/\.less(?:\?.*)?$/.test(req.url) && contentType === 'application/octet-stream')) {
                    function sendErrorResponse(errorMessage, cssText) {
                        errorMessage = "express-compiless: Error compiling " + req.originalUrl + ":\n" + errorMessage;
                        res.removeHeader('Content-Length');
                        res.setHeader('Content-Type', 'text/css');
                        res.send("body:before {display: block; white-space: pre; font-family: 'Courier New', monospace; font-size: 24px; color: black; margin: 10px; padding: 10px; border: 4px dashed red; margin-bottom: 10px; content: '" + errorMessage.replace(/'/g, "\\'").replace(/\n/g, '\\00000a') + "'}\n" + (cssText || ''));
                    }

                    // If there's an ETag, make sure it's different from the original one so the downstream middleware
                    // won't reply with a false positive 304 on the same URL after compiless has been enabled or disabled:
                    var chunks = [];
                    res.on('error', function () {
                        res.unhijack();
                        next();
                    }).on('data', function (chunk) {
                        chunks.push(chunk);
                    }).on('end', function () {
                        if (!chunks.length) {
                            return res.send(res.statusCode);
                        }
                        var lessText = Buffer.concat(chunks).toString('utf-8'), // No other charsets are really relevant, right?
                            baseDir = Path.resolve(options.root, req.url.replace(/\/[^\/]*(?:\?.*)?$/, '/').substr(1)),
                            isSeenByFileName = {},
                            errors = [];

                        flattenLessText(lessText, baseDir, isSeenByFileName, errors, function (err, flattenedLessText) {
                            if (err) {
                                errors.push(err);
                            }

                            seq(Object.keys(isSeenByFileName))
                                .parEach(function (fileName) {
                                    fs.stat(fileName, this.into(fileName));
                                })
                                .unflatten()
                                .seq(function (fileNames) {
                                    var oldETag = res.getHeader('ETag');
                                    if (oldETag) {
                                        var etagFragments = [oldETag.replace(/^"|"$/g, '')];
                                        if (fileNames.length) {
                                            var importedFileStats = [];
                                            fileNames.sort().forEach(function (fileName) {
                                                var stats = this.vars[fileName];
                                                importedFileStats.push(fileName, String(stats.mtime.getTime()), String(stats.size));
                                            }, this);
                                            etagFragments.push(crypto.createHash('md5').update(importedFileStats.join('-')).digest('hex').substr(0, 16));
                                        }
                                        var newETag = '"' + etagFragments.join('-') + '-compiless"';
                                        res.setHeader('ETag', newETag);

                                        if (ifNoneMatch && ifNoneMatch.indexOf(newETag) !== -1) {
                                            return res.send(304);
                                        }
                                    }

                                    // Unfortunately less.render throws errors in async code instead of passing it to our callback.
                                    // Hopefully the solution for https://github.com/cloudhead/less.js/issues/462 will remedy this.
                                    less.render(flattenedLessText, {paths: [baseDir]}, function (err, cssText) {
                                        if (err) {
                                            errors.push(err);
                                        }
                                        if (errors.length) {
                                            return sendErrorResponse(_.pluck(errors, 'message').join("\n"), cssText);
                                        }
                                        cssText = fileNames.map(function (fileName) {
                                            return ".compilessinclude {background-image: url(" + Path.relative(baseDir, fileName) + "); display: none;}\n";
                                        }).join("") + cssText;
                                        res.setHeader('Content-Type', 'text/css');
                                        res.setHeader('Content-Length', Buffer.byteLength(cssText));
                                        res.end(cssText);
                                    });
                               });
                        });
                    });
                } else {
                    res.unhijack(true);
                }
            });
            next();
        } else {
            next();
        }
    };
};


lib/flattenLessText.js
var Path = require('path'),
    fs = require('fs'),
    seq = require('seq'),
    passError = require('passerror'),
    urlTokenRegExp = /url\((['"]|)(.*?)\1\)/g,
    importRegExp = /(@import(?:\s*'.*?'|\s*".*?"|\s+url\((?:[^'"]*?|'[^']*'|"[^"]*")\));)/,
    importRegExpWithCaptures = /@import(?:\s*(['"])(.*?)\1|\s+url\((['"]|)(.*?)\3\));/;

// Returns flattened less text, accumulates seen files in the isSeenByFileName object provided by the caller.
module.exports = function flattenLessText(lessText, flattenedLessTextDir, baseDir, isSeenByFileName, errors, cb) {
    // Remove @-signs in comments so we don't accidentally resolve commented-out imports:
    lessText = lessText.replace(/\/\*([\s\S]*?)\*\//g, function ($0, commentBody) {
        return '/*' + commentBody.replace(/@/g, ' ') + '*/';
    }).replace(/\/\/([^\r\n]*)/g, function ($0, commentBody) {
        return '//' + commentBody.replace(/@/g, ' ');
    });
    seq(lessText.split(importRegExp))
        .parMap(function (chunk) {
            var cb = this,
                matchImport = chunk.match(importRegExpWithCaptures);
            if (matchImport) {
                var importedFileName = Path.resolve(baseDir.replace(/\/*$/, '/'), matchImport[2] || matchImport[4]),
                    importedFileDir = Path.dirname(importedFileName);
                fs.readFile(importedFileName, 'utf-8', function (err, importedLessText) {
                    if (err) {
                        errors.push(err);
                        cb(null, '');
                    } else {
                        isSeenByFileName[importedFileName] = true;
                        flattenLessText(importedLessText, flattenedLessTextDir, importedFileDir, isSeenByFileName, errors, passError(cb, function (lessText) {
                            lessText = lessText.replace(urlTokenRegExp, function ($0, quoteChar, href) {
                                if (/^\w+:/.test(href) || /^\//.test(href) || /@/.test(href)) { // Hack: Skip url(...) that use macro expansion
                                    // Skip absolute, protocol-relative, and root-relative urls
                                    return $0;
                                } else {
                                    // Make the href relative to the directory of the flattened file:
                                    var referencedFileName = Path.resolve(importedFileDir, href),
                                        rewrittenHref = Path.relative(flattenedLessTextDir + '/', referencedFileName);

                                    // Quote if necessary:
                                    if (/^[a-z0-9\/\-_.]*$/i.test(rewrittenHref)) {
                                        return "url(" + rewrittenHref + ")";
                                    } else {
                                        return "url('" + rewrittenHref.replace(/([\'\"])/g, "\\$1") + "')";
                                    }
                                }
                            });
                            cb(null, lessText);
                        }));
                    }
                });
            } else {
                cb(null, chunk);
            }
        })
        .catch(function (err) {
            // Why doesn't this work?!
            cb(err);
        })
        .seq(function () {
            cb(null, this.stack.join(''));
        });
}


lib/compiless.js
const Path = require('path');
const crypto = require('crypto');
const fs = require('fs');
const async = require('async');
const csserror = require('csserror');
const passError = require('passerror');
const hijackResponse = require('hijackresponse');

module.exports = function compiless(options) {
  if (!options || !options.root) {
    throw new Error('options.root is mandatory');
  }

  const less = options.less || require('less');
  const isOldLessApi = !less.version || less.version[0] < 2;
  const plugins = options.plugins || [];
  if (plugins.length > 0 && isOldLessApi) {
    throw new Error('Less version < 2.0.0 does not support plugins');
  }
  const toCSSOptions = options.toCSSOptions || {};

  function formatError(err) {
    // err can either be a string or an error object
    if (typeof err === 'string') {
      return err;
    } else {
      // Assume error object
      return (
        err.message +
        ('line' in err
          ? ' at line ' +
            err.line +
            ('column' in err ? ', column ' + err.column : '')
          : '') +
        ('extract' in err ? ':\n' + err.extract.join('\n') : '')
      );
    }
  }

  return (req, res, next) => {
    // Prevent If-None-Match revalidation with the downstream middleware with ETags that aren't suffixed with "-compiless":
    const ifNoneMatch = req.headers['if-none-match'];

    // only hijack requests for .less files
    if (/\.less$/.test(req.originalUrl)) {
      if (ifNoneMatch) {
        const validIfNoneMatchTokens = ifNoneMatch
          .split(' ')
          .filter((etag) => /-compiless"$/.test(etag));

        if (validIfNoneMatchTokens.length > 0) {
          req.headers['if-none-match'] = validIfNoneMatchTokens.join(' ');
        } else {
          delete req.headers['if-none-match'];
        }
      }
      delete req.headers['if-modified-since']; // Prevent false positive conditional GETs after enabling compiless

      return hijackResponse(res, next).then(({ readable, writable }) => {
        const contentType = res.getHeader('Content-Type');
        const matchContentType =
          contentType &&
          contentType.match(/^text\/less(?:;\s*charset=([a-z0-9-]+))?$/i);

        // The mime module doesn't support less yet, so we fall back:
        if (!(matchContentType || contentType === 'application/octet-stream')) {
          return readable.pipe(writable);
        }

        const baseDir = Path.resolve(
          options.root,
          req.url.replace(/\/[^/]*(?:\?.*)?$/, '/').substr(1)
        );
        const filename = Path.resolve(options.root, req.url.substr(1));
        const lessOptions = {
          paths: [baseDir],
          filename,
          plugins,
          relativeUrls: true,
        };

        function sendErrorResponse(err, cssText) {
          const errorMessage =
            'express-compiless: Error compiling ' +
            req.originalUrl +
            ':\n' +
            formatError(err);
          res.removeHeader('Content-Length');
          res.removeHeader('ETag');
          res.setHeader('Content-Type', 'text/css; charset=utf-8');
          writable.end(csserror(errorMessage) + '\n' + (cssText || ''));
        }

        function respondWithCss(css, importedFileNames) {
          importedFileNames = importedFileNames || [];
          const statsByFileName = {};
          async.eachLimit(
            importedFileNames,
            10,
            (importedFileName, cb) => {
              fs.stat(
                importedFileName,
                passError(cb, (stats) => {
                  statsByFileName[importedFileName] = stats;
                  cb();
                })
              );
            },
            passError(sendErrorResponse, function () {
              const oldETag = res.getHeader('ETag');

              if (oldETag) {
                const oldETagIsWeak = oldETag && /^W\//.test(oldETag);
                const etagFragments = [oldETag.replace(/^(?:W\/)?"|"$/g, '')];

                if (importedFileNames.length) {
                  const importedFileStats = [];

                  importedFileNames.sort().forEach((importedFileName) => {
                    const stats = statsByFileName[importedFileName];
                    importedFileStats.push(
                      importedFileName,
                      String(stats.mtime.getTime()),
                      String(stats.size)
                    );
                  }, this);

                  etagFragments.push(
                    crypto
                      .createHash('md5')
                      .update(importedFileStats.join('-'))
                      .digest('hex')
                      .substr(0, 16)
                  );
                }

                const newETag =
                  (oldETagIsWeak ? 'W/' : '') +
                  '"' +
                  etagFragments.join('-') +
                  '-compiless"';
                res.setHeader('ETag', newETag);

                if (ifNoneMatch && ifNoneMatch.indexOf(newETag) !== -1) {
                  res.status(304);
                  return writable.end();
                }
              }

              const cssText =
                importedFileNames
                  .map(
                    (importedFileName) =>
                      '.compilessinclude {background-image: url(' +
                      Path.relative(baseDir, importedFileName) +
                      '); display: none;}\n'
                  )
                  .join('') + css;

              res.setHeader('Content-Type', 'text/css; charset=utf-8');
              res.setHeader('Content-Length', Buffer.byteLength(cssText));
              writable.end(cssText);
            })
          );
        }

        const chunks = [];
        readable
          .on('data', (chunk) => {
            chunks.push(chunk);
          })
          .on('end', () => {
            const lessText = Buffer.concat(chunks).toString();

            if (isOldLessApi) {
              const parser = new less.Parser(lessOptions);
              parser.parse(
                lessText,
                passError(sendErrorResponse, (root) => {
                  let css;
                  try {
                    css = root.toCSS(toCSSOptions);
                  } catch (e) {
                    return sendErrorResponse(e);
                  }
                  respondWithCss(
                    css,
                    parser.imports &&
                      parser.imports.files &&
                      Object.keys(parser.imports.files)
                  );
                })
              );
            } else {
              less.render(
                lessText,
                lessOptions,
                passError(sendErrorResponse, (output) => {
                  respondWithCss(output.css, output.imports);
                })
              );
            }
          });
      });
    }
    next();
  };
};
